### ISSUE TITLE
Deadkeys that are supposed to prevent matches are not working in KeymanWeb

### ISSUE THREAD

**Issue by darcywong00** [keymanapp/keyman#281]
<a href="https://github.com/mcdurdin"><img src="https://avatars1.githubusercontent.com/u/4498365?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [mcdurdin](https://github.com/mcdurdin)**
_Wednesday Oct 9, 2013 at 22:31 GMT_
_Originally opened as KD-157_

----

When a deadkey is in the middle of a context string, a context should not match if it does not reference that deadkey. For example, if the backing store is:
```
  a<dk(0)>b
```
Then:
```
  'ab' + 'c' > 'match'
```
Should not fire. This is the Keyman Engine behaviour. The reason the rule is firing is that we don't ever test deadkeys within the context match rules.

## Keyboard which replicates this bug in KeymanWeb 
```
store(&VERSION) '8.0'
store(&NAME) 'Test missing deadkey'
begin Unicode > use(main)

group(main) using keys

'ab' + 'c' > 'match'
+ 'b' > dk(1) 'b'

+ 'd' > 'd' dk(2)
'd' + 'e' > 'fail'
```

## Test cases

**Type:** "abc".
**Expected (Keyman Engine):** "a<dk>bc"
**Actual (KeymanWeb):** "match"

**Type** "de".
**Expected (Keyman Engine):** "d<dk>e"
**Actual (KeymanWeb):** "fail"

I do not have an immediate answer for a fix. It will probably require a compiler change, because updating KCM to test for deadkeys within the text will work, but testing for a deadkey at the end of the text is not going to work. This is due to the way the compiler assembles the tests, e.g.
```
  'd' dk(2) + 'E' > 'pass'
```
Compiles to:
```
  if(k.KKM(e,16400,69)&&k.KCM(1,t,"d",1)&&k.KDM(0,t,1))
```
But:
```
  'd' + 'e' > 'fail'     
```
Compiles to:
```
  if(k.KKM(e,16384,69)&&k.KCM(1,t,"d",1))
```

So we need to have some way of validating the deadkey state even for rules that don't specifically test for a deadkey.

----
MD and JD add:

One keyboard that exhibits this issue is ISIS Devangari, typing 'kxi' or 'kxh'.
We expect कइ and कह respectively. But we get the default form of कि and ख instead.

This behaviour was confirmed on 2 May 2014.
If it is a Keyman compiler bug, then the issue should be moved to KM Dev. But if it needs changes to the KMEW code, then it should stay here.

**Comment by jahorton** [comment]
This behavior is confirmed to still exist as of October 13, 2017 when using ISIS Devangari with keymanweb.com?tier=alpha.

**Comment by jahorton** [comment]
My recent re-investigation of the issue suggests that work will need to be done in both KMW and Developer to achieve a proper fix moving forward with 10.0 while maintaining compatibility with earlier versions.

**Comment by jahorton** [comment]
More work is still needed on KeymanWeb and Developer to fully fix this issue; one component PR was mislabled as fixing the issue.
